openpype/hosts/maya/plugins/publish/extract_xgen.py

```diff
@@ -0,0 +1,139 @@
+import os
+import copy
+
+from maya import cmds
+import pymel.core as pc
+import xgenm
+
+from openpype.pipeline import publish
+from openpype.hosts.maya.api.lib import maintained_selection, attribute_values
+from openpype.lib import StringTemplate
+
+
+class ExtractXgenCache(publish.Extractor):
+    """Extract Xgen"""
+
+    label = "Extract Xgen"
+    hosts = ["maya"]
+    families = ["xgen"]
+    scene_type = "ma"
+
+    def process(self, instance):
+        if "representations" not in instance.data:
+            instance.data["representations"] = []
+
+        staging_dir = self.staging_dir(instance)
+        maya_filename = "{}.{}".format(instance.data["name"], self.scene_type)
+        maya_filepath = os.path.join(staging_dir, maya_filename)
+
+        # Get published xgen file name.
+        template_data = copy.deepcopy(instance.data["anatomyData"])
+        template_data.update({"ext": "xgen"})
+        templates = instance.context.data["anatomy"].templates["publish"]
+        xgen_filename = StringTemplate(templates["file"]).format(template_data)
+        name = instance.data["xgenPalette"].replace(":", "__").replace("|", "")
+        xgen_filename = xgen_filename.replace(".xgen", "__" + name + ".xgen")
+
+        # Export xgen palette files.
+        xgen_path = os.path.join(staging_dir, xgen_filename).replace("\\", "/")
+        xgenm.exportPalette(
+            instance.data["xgenPalette"].replace("|", ""), xgen_path
+        )
+        self.log.info("Extracted to {}".format(xgen_path))
+
+        representation = {
+            "name": name,
+            "ext": "xgen",
+            "files": xgen_filename,
+            "stagingDir": staging_dir,
+        }
+        instance.data["representations"].append(representation)
+
+        # Collect nodes to export.
+        duplicate_nodes = []
+        for node, connections in instance.data["xgenConnections"].items():
+            transform_name = connections["transform"].split(".")[0]
+
+            # Duplicate_transform subd patch geometry.
+            duplicate_transform = pc.duplicate(transform_name)[0]
+            pc.parent(duplicate_transform, world=True)
+            duplicate_transform.name(stripNamespace=True)
```

Comment:
Could we include in a comment why there is a need to duplicate the input transforms, etc? Why do we need that. Couldn't we just export a maya file with `preserveReferences = false` since duplicating breaks the reference connection anyway? What's the need for duplicating it?

Reply:
We need to strip out any connections the geometry has to the current scene. For example when using some rig geometry that is nested deep in the rig.
Up for suggestions on how to solve this without duplicating and parent to the world.

Reply:
> We need to strip out any connections the geometry has to the current scene. For example when using some rig geometry that is nested deep in the rig.
Up for suggestions on how to solve this without duplicating and parent to the world.

I see, thanks. I believe you should be able to add these arguments to the cmds file export `channels=False` and `constructionHistory=False`. However, whenever you do so any connections you want to actively persist in the export would need to be included in the "selection" that you've set to export.

So basically if you're able to identify which actual objects (including its history) you want to have end up in the maya file you should be able to add those arguments (make sure `preserveReference=False` otherwise any reference in your selection would still fully be included).

I believe if the selection actively includes animation curves (if you ever need them) then `channels=False` and `constructionHistory=False` will still include it in the export and connect them as they were as long as they are part of the active selection in an `exportSelected=True` export.

However, you might need to double test this. :)

Reply:
But this includes the namespaces of nodes, right?

Reply:
> But this includes the namespaces of nodes, right?

Yes, you are correct. It doesn't remove namespaces as far as I'm aware.